samples/js/photos/src/main.js

```diff
@@ -0,0 +1,79 @@
+// @flow
+
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+import React from 'react';
+import ReactDOM from 'react-dom';
+import {searchToParams} from './dom.js';
+import Nav from './nav.js';
+import PhotosPage from './photos-page.js';
+import Viewport from './viewport.js';
+import {createPhoto} from './photo.js';
+import type {PhotoIndex, NomsPhoto} from './types.js';
+import {Path, PathSpec, Struct} from '@attic/noms';
+
+// Cache of index paths to indices. Otherwise calls to render are pretty slow,
+// which is noticeable when resizing, toggling between full screen photos, etc.
+const indexMap: Map<string, PhotoIndex> = new Map();
```

Comment:
Mm I messed up here. The answer is actually no, there is no in-memory cache, and that patch won't help.

The problem is that the spec parsing code recreates the database each time, based on the spec. There is no way to re-use a database.

Solution 1: a similar trick in the JS code that we use for leveldb, and re-use the databases. However, this is kind of icky because it means subsequent calls to database() would ignore the cache size parameter. I'd also need to do a bit more work to thread through the cacheSize parameter through the value() call, which I missed.

Solution 2: a factory pattern: refactor spec parsing into a SpecParser class which is constructed with a cacheSize parameter, then has parsePath, parseDataset, and parseDatabase methods which pass the correct cacheSize parameter.

@aboodman @arv but right now, I think we should leave this cache in place and just submit this PR.